dipy/direction/ptt_direction_getter.pyx

```diff
@@ -0,0 +1,472 @@
+                              + self.propagator[5] * frame[2][i])
+            normalize(tangent)
+
+            if q < (self.probe_quality - 1):
+                for i in range(3):
+                    binormal[i] = (self.propagator[6] * frame[0][i]
+                                   + self.propagator[7] * frame[1][i]
+                                   + self.propagator[8] * frame[2][i])
+                cross(normal, binormal, tangent)
+                for i in range(3):
+                    frame[0][i] = tangent[i]
+                    frame[1][i] = normal[i]
+                    frame[2][i] = binormal[i]
+
+            if self.probe_count == 1:
+                fod_amp = self.pmf_gen.get_pmf_value(position, tangent)
+                fod_amp = fod_amp if fod_amp > self.pmf_threshold else 0
+                self.last_val_cand = fod_amp
+                likelihood += self.last_val_cand
+            else:
+                self.last_val_cand = 0
+                if q == self.probe_quality-1:
+                    for i in range(3):
+                        binormal[i] = (self.propagator[6] * frame[0][i]
+                                      + self.propagator[7] * frame[1][i]
+                                      + self.propagator[8] * frame[2][i])
+                    cross(normal, binormal, tangent)
+
+                for c in range(self.probe_count):
+                    for i in range(3):
+                        new_position[i] = (position[i]
+                                           + normal[i] * self.probe_radius
+                                           * cos(c * self.angular_separation)
+                                           + binormal[i] * self.probe_radius
+                                           * sin(c * self.angular_separation))
+                    fod_amp = self.pmf_gen.get_pmf_value(new_position, tangent)
+                    fod_amp = fod_amp if fod_amp > self.pmf_threshold else 0
+                    self.last_val_cand += fod_amp
+
+                likelihood += self.last_val_cand
+
+        likelihood *= self.probe_normalizer
+        if self.data_support_exponent != 1:
+            likelihood = pow(likelihood, self.data_support_exponent)
+
+        return likelihood
+
+
+    cdef int initialize(self, double[:] seed_point, double[:] seed_direction):
+        """Sample an initial curve by rejection sampling.
+
+        Parameters
+        ----------
+        seed_point : double[3]
+            Initial point
+        seed_direction : double[3]
+            Initial direction
+
+        Returns
+        -------
+        status : int
+            Returns 0 if the initialization was successful, or
+            1 otherwise.
+        """
+
+        self.position[0] = seed_point[0]
+        self.position[1] = seed_point[1]
+        self.position[2] = seed_point[2]
+
+        cdef double data_support = 0
+        cdef double max_posterior = 0
+
+        cdef int tries
+        for tries in range(1000):
```

Comment:
1000 could be a variable if we want to reduces tries number. 

Also, the loop is never break. Does it means that it will do the 1000 tries even if max_posterior is good ?

Reply:
Thanks. I added the variable. It should have been 20, as in the `propagate(.)` function. The loop never breaks because it is estimating the maximum likelihood for the rejection sampling algorithm.

Reply:
This loop estimates the max_posterior. The estimation is done by picking the maximum data_support among 1000 random trials. So all 1000 trials are used that is why the loop should not break. Here, 1000 is quite an arbitrary choice. In my opinion, this should be sufficient but we have not thoroughly investigated this. 20 on the other hand sounds very small in my opinion, it is quite likely that the max_posterior will be underestimated in that case. Maybe, for now we can stick to a high value, like 1000, to be on the safe side.

Reply:
20 might be sufficient during propagation though, because the tangent is fixed. But at the initialization, the tangent is also most of the times not fixed. So if it is the case that there is an initial direction, then 20 can be used instead of 1000.

Reply:
Thanks for the clarification @baranaydogan. The rejection sampling is suboptimally implemented now and can be improved. I set the `tries` to 20 (now 10) into the initialization loop, as in the dipy implementation the initial tangent is always set to one of the peaks at the seed location. This is also the case for probabilistic DG. This is something to generalize in a future the tracking module overhaul.

Reply:
Things differ in that case of course. This is good know. Thank you for the information.